File: packages/guardian-rn/ios/Sources/GuardianRN/HmacSigner.swift
import Foundation
import CommonCrypto

/// Signs a canonical-JSON payload with HMAC-SHA256 (ADR-0003).
/// Output format: "sha256=<lowercase hex>"
enum HmacSigner {

    static func sign(canonicalPayload: String, keyBytes: [UInt8]) -> String {
        var digest = [UInt8](repeating: 0, count: Int(CC_SHA256_DIGEST_LENGTH))
        let data = Array(canonicalPayload.utf8)
        CCHmac(CCHmacAlgorithm(kCCHmacAlgSHA256),
               keyBytes, keyBytes.count,
               data, data.count,
               &digest)
        let hex = digest.map { String(format: "%02x", $0) }.joined()
        return "sha256=\(hex)"
    }

    /// Constant-time comparison to prevent timing side-channels.
    static func verify(canonicalPayload: String, keyBytes: [UInt8], expected: String) -> Bool {
        let computed = sign(canonicalPayload: canonicalPayload, keyBytes: keyBytes)
        // expected is the untrusted user input, computed is the trusted string

        var expectedIterator = expected.utf8.makeIterator()

        var diff: Int = computed.utf8.count ^ expected.utf8.count

        for aChar in computed.utf8 {
            let bChar = expectedIterator.next() ?? 0
            diff |= Int(aChar) ^ Int(bChar)
        }
        return diff == 0
    }
}

File: packages/guardian-rn/android/src/main/kotlin/com/guardian/rn/HmacSigner.kt
package com.guardian.rn

import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

/**
 * Signs a canonical-JSON payload with HMAC-SHA256.
 * Per ADR-0003: envelope.hmac = "sha256=<hex(HMAC-SHA256(canonicalJson(payload), sessionKey))>"
 */
object HmacSigner {

    fun sign(canonicalPayload: String, keyBytes: ByteArray): String {
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(keyBytes, "HmacSHA256"))
        val digest = mac.doFinal(canonicalPayload.toByteArray(Charsets.UTF_8))
        return "sha256=" + digest.joinToString("") { "%02x".format(it) }
    }

    fun verify(canonicalPayload: String, keyBytes: ByteArray, expectedHmac: String): Boolean {
        val computed = sign(canonicalPayload, keyBytes)
        // Constant-time comparison to prevent timing attacks
        return constantTimeEquals(computed, expectedHmac)
    }

    private fun constantTimeEquals(trusted: String, untrusted: String): Boolean {
        var diff = trusted.length xor untrusted.length
        for (i in trusted.indices) {
            val untrustedCode = if (i < untrusted.length) untrusted[i].code else 0
            diff = diff or (trusted[i].code xor untrustedCode)
        }
        return diff == 0
    }
}
